include/cantera/zeroD/ReactorDelegator.h

```diff
@@ -0,0 +1,381 @@
+//! can be accessed by casting a Reactor* to a ReactorAccessor* without needing
+//! to know the specific kind of Reactor at compilation time.
+class ReactorAccessor
+{
+public:
+    //! Set the number of equations represented by this reactor
+    virtual void setNEq(size_t n) = 0;
+
+    //! Get the net rate of volume change (for example, from moving walls) [m^3/s]
+    virtual double vdot() const = 0;
+
+    //! Set the net rate of volume change (for example, from moving walls) [m^3/s]
+    virtual void setVdot(double v) = 0;
+
+    //! Get the net heat transfer rate (for example, through walls) [W]
+    virtual double qdot() const = 0;
+
+    //! Set the net heat transfer rate (for example, through walls) [W]
+    virtual void setQdot(double q) = 0;
+
+    //! Set the state of the thermo object to correspond to the state of the reactor
+    virtual void restoreThermoState() = 0;
+
+    //! Set the state of the thermo object for surface *n* to correspond to the
+    //! state of that surface
+    virtual void restoreSurfaceState(size_t n) = 0;
+};
+
+//! Delegate methods of the Reactor class to external functions
+template <class R>
+class ReactorDelegator : public Delegator, public R, public ReactorAccessor
+{
+public:
+    ReactorDelegator() {
+        m_initialize = [this](double t0) { R::initialize(t0); };
+        m_syncState = [this]() { R::syncState(); };
+        m_getState = [this](double* y) { R::getState(y); };
+        m_updateState = [this](double* y) { R::updateState(y); };
+        m_updateSurfaceState = [this](double* y) { R::updateSurfaceState(y); };
+        m_getSurfaceInitialConditions = [this](double* y) {
+            R::getSurfaceInitialConditions(y);
+        };
+        m_updateConnected = [this](bool updatePressure) {
+            R::updateConnected(updatePressure);
+        };
+        m_eval = [this](double t, double* ydot) { R::eval(t, ydot); };
+        m_evalWalls = [this](double t) { R::evalWalls(t); };
+        m_evalSurfaces = [this](double t, double* ydot) {
+            return R::evalSurfaces(t, ydot);
+        };
+        m_componentName = [this](size_t k) { return R::componentName(k); };
+        m_componentIndex = [this](const std::string& nm) {
+            return R::componentIndex(nm);
+        };
+        m_speciesIndex = [this](const std::string& nm) {
+            return R::speciesIndex(nm);
+        };
+    }
+
+    virtual void setDelegate(const std::string& name,
+                             const std::function<void()>& func,
+                             const std::string& when) override
+    {
+        if (name == "syncState") {
+            m_syncState = makeDelegate(func, when,
+                [this]() { R::syncState(); }
+            );
+        } else {
+            throw NotImplementedError("ReactorDelegator::setDelegate",
+                "For function named '{}' with signature void(double)", name);
+        }
+    }
+
+    virtual void setDelegate(const std::string& name,
```

Comment:
I see that much of the boilerplate of the `ReactorDelegator` is to define `setDelegate` for various lambda function signatures. Would it be possible to define a new class that handles various lambda functions in a more generic fashion? I am asking as this would become highly beneficial if the `Delegator` concept would become more widespread. 

That way, `setDelegate` may hopefully simplify to
```C++
    virtual void setDelegate(const std::string& name,
                             const GenericFunc& func,
                             const std::string& when) override
```
with `makeDelegate` becoming a member function of `GenericFunc`. In that way, all cases could potentially be handled by a single `if` - `else if` cascade?

I am also asking as I just experimented with the old `Func1` objects this morning, which are quite inflexible. Expanding this to a `GenericFunc` concept (potentially with overloaded specializations) would presumably have a range of other applications.

Reply:
> **PS:** I apologize for adding this specific comment about the implementation early; I was curious after running into a `Func1`-related roadblock earlier today.

FWIW, I think this is also the purpose of draft pull requests, to identify potentially large issues before they get committed or before too much effort is expended and have to be changed.

Reply:
I think the challenge is that a function with an unspecified signature is an incomplete type, so you'd have to define child classes of `GenericFunc` for each function signature, with a member variable holding a `std::function` of the specific type. Then, you'd have to check for each option in the single `setDelegate` function and do a `dynamic_cast` to be able to get at the underlying `std::function` object in order to be able set the appropriate member variable. So I think the current implementation is preferable.

As explained in Cantera/enhancements#79, the intention is absolutely to employ this infrastructure pretty widely in Cantera - I'm envisioning it being used for phase thermo, species thermo, kinetics, reactions, transport, reactors, and 1D systems. I agree that there is a little more boilerplate here than one might hope, but I think I've already worked out most of the important function patterns that we need to support, and most of what's needed in the future should follow directly from this.

Reply:
While I am still partial to having a `GenericFunction` defined outside of delegated methods, some of my concerns here boil down to relying on `if`-`else` cascades to detect what is done where. 

I believe much of this would be more readable (and likely more maintainable) if it were implemented as a series of class methods. I realize that functions need to be recognized by `name`, but that could be handled by a `std::map`?